randovania/generator/dock_weakness_distributor.py

```diff
@@ -0,0 +1,193 @@
+    """
+
+    descriptions = {
+        player: default_database.game_description_for(patches.configuration.game)
+        for player, patches in all_patches.items()
+    }
+
+    unassigned_docks: list[Tuple[int, NodeIdentifier]] = []
+
+    for player, patches in all_patches.items():
+        player_docks: list[Tuple[int, NodeIdentifier]] = []
+
+        if patches.configuration.dock_rando.mode == DockRandoMode.ONE_WAY:
+            player_docks.extend(((player, identifier) for identifier in patches.dock_weakness.keys()))
+        
+        if patches.configuration.dock_rando.mode == DockRandoMode.TWO_WAY:
+            game = descriptions[player]
+            ctx = NodeContext(
+                patches,
+                patches.starting_items,
+                game.resource_database,
+                game.world_list
+            )
+
+            for identifier in patches.dock_weakness:
+                dock: DockNode = game.world_list.node_by_identifier(identifier)
+                if (player, dock.get_target_identifier(ctx)) not in player_docks:
+                    player_docks.append((player, identifier))
+        
+        if TO_SHUFFLE_PROPORTION < 1.0:
+            rng.shuffle(player_docks)
+            limit = int(len(player_docks) * TO_SHUFFLE_PROPORTION)
+            player_docks = player_docks[:limit]
+        
+        unassigned_docks.extend(player_docks)
+    
+    rng.shuffle(unassigned_docks)
+
+    docks_placed = 0
+    docks_to_place = len(unassigned_docks)
+
+    start_time = time.perf_counter()
+
+    while unassigned_docks:
+        await asyncio.sleep(0)
+        status_update(f"{docks_placed}/{docks_to_place} docks placed")
+
+        player, identifier = unassigned_docks.pop()
+
+        game = descriptions[player]
+        patches = all_patches[player]
+
+        dock: DockNode = game.world_list.node_by_identifier(identifier)
+        target: DockNode = game.world_list.node_by_identifier(dock.get_target_identifier(NodeContext(
+            patches,
+            patches.starting_items,
+            game.resource_database,
+            game.world_list,
+        )))
+
+        dock_type_params = game.dock_weakness_database.dock_rando_params[dock.dock_type]
+        dock_type_state = patches.configuration.dock_rando.types_state[dock.dock_type]
+
+        # Calculate the reach assuming this dock is completely locked
+        locks = {identifier: dock_type_params.locked}
+        if patches.configuration.dock_rando.mode == DockRandoMode.TWO_WAY:
+            locks[target.identifier] = dock_type_params.locked
+        locked_patches = patches.assign_dock_weakness_assignment(locks)
+
+        state, logic = resolver.setup_resolver(patches.configuration, locked_patches)
+        logic = DockRandoLogic.from_logic(logic, dock)
+        debug.log_resolve_start()
+        debug.debug_print(f"{identifier}")
+        
+        level = debug.debug_level()
+        debug.set_level(0)
```

Comment:
Could have something like this in `debug.py`:
```python
@contextlib.contextmanager
def with_level(level: int):
    current_level = debug_level()
    set_level(level)
    yield
    set_level(current_level)
```

So here you'd just have:
```python
with debug.with_level(0):
    try:
        new_state = await resolver.advance_depth(state, logic, lambda s: None, max_attempts=RESOLVER_ATTEMPTS)
    except resolver.ResolverTimeout:
        new_state = None
    # result stuff
```

Reply:
:+1: